### ISSUE TITLE
CodepipelineStageActions InputArtifacts & OutputArtifacts validation does not work with Intrinsic Functions

### ISSUE THREAD

**Issue by bensoer** [aws-cloudformation/cfn-lint#2138]
**Versions**
*cfn-lint version: 0.54.2*
*python version: 3.6*

**Description**
My work is using cfn-lint for validating their CloudFormation template files. These template files include the creation of complex CodePipelines with multiple stages. Each stages' `OutputArtifacts` and `InputArtifacts` also use `Fn::If` conditions which specify what artifacts are available depending, in my work's case, the programming language that is being used for the project the CodePipeline is for. 

cfn-lint does not parse these `Fn::If` statements and thus fails to validate the `InputArtifacts` and `OutputArtifacts` correctly. It prints out this error:
```
E2541 Every input artifact for an action must match the output artifact of an action earlier in the pipeline. (cfn_templates)
cfn/dist/resource.yaml:1132:15

E2541 Every input artifact for an action must match the output artifact of an action earlier in the pipeline. (deploy_templates)
cfn/dist/resource.yaml:1239:19
```
Eventhough the CloudFormation template works.

Looking at the code here, the parsing seems pretty rudimentary, with no checks or consideration for AWS Intrinsic functions: https://github.com/aws-cloudformation/cfn-lint/blob/34affe23fde9526b50f0800f2fdbb83d6c6bab1b/src/cfnlint/rules/resources/codepipeline/CodepipelineStageActions.py#L223

A similar issue was reported with conditionals but it seems the changes is not something global that is being used in CodePipeline: https://github.com/aws-cloudformation/cfn-lint/issues/350

**Sample CloudFormation**
A sample piece of our CloudFormation `OutputArtifacts` of the template looks like this:
```yaml
OutputArtifacts:
    'Fn::If':
        - EnableProjectAssetsPublish
        - 'Fn::If':
            - EnableOpenapiSpecPublish
            - 'Fn::If':
                - ProjectPublishesCode
                - - Name: openapi_specs
                   - Name:
                        'Fn::If':
                            - AutoCondition0
                            - maven_jars
                            - 'Fn::If':
                                - AutoCondition1
                                - npm_packages
                                - 'Fn::If':
                                    - AutoCondition2
                                    - py_packages
                                    - Ref: 'AWS::NoValue'
                   - Name: cfn_templates
                  - Name: deploy_templates
                  - Name: project_assets
                - - Name: openapi_specs
                   - Name: cfn_templates
                   - Name: deploy_templates
                   - Name: project_assets
```

**Current Work Around**
Currently we disable rule `E2541` but this does defeat the purpose of a lot of our templates which contain complex pipelines

**AWS References**
- https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html
- https://docs.aws.amazon.com/codepipeline/latest/userguide/welcome-introducing-artifacts.html
